src/tests/Stream/PartitionFilterTest.cpp

```diff
@@ -0,0 +1,165 @@
+//
+// See LICENSE.md at the top level for license information.
+//
+// Screaming Udder!                              https://esss.se
+
+/// \brief Test partition filtering.
+///
+
+#include "Stream/PartitionFilter.h"
+#include <chrono>
+#include <gtest/gtest.h>
+#include <thread>
+
+using std::chrono_literals::operator""ms;
+using Stream::time_point;
+class PartitionFilterTest : public ::testing::Test {
+public:
+  Stream::PartitionFilter UnderTest{Stream::time_point::max(), 10ms, 20ms};
+};
+
+TEST_F(PartitionFilterTest, InitState) {
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, MessageNoStop) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Message));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateNoStop) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_TRUE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateNoStopAlt) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_TRUE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, TimeoutNoStop) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::TimedOut));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, EmptyNoStop) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::TimedOut));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecovered) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Message));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecoveredAlt) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  std::this_thread::sleep_for(40ms);
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Message));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecoveredEOP) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_FALSE(
+      UnderTest.shouldStopPartition(KafkaW::PollStatus::EndOfPartition));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecoveredEOPAlt) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  std::this_thread::sleep_for(40ms);
+  EXPECT_FALSE(
+      UnderTest.shouldStopPartition(KafkaW::PollStatus::EndOfPartition));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecoveredEmpty) {
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Error));
+  EXPECT_FALSE(UnderTest.shouldStopPartition(KafkaW::PollStatus::Empty));
+  EXPECT_FALSE(UnderTest.hasErrorState());
+}
+
+TEST_F(PartitionFilterTest, ErrorStateRecoveredEmptyAlt) {
```

Comment:
This test and `ErrorStateRecoveredTimeOutAlt` look superfluous to me.

Reply:
I am impressed that you found this and you are (in my opinion) kind of correct. Both tests executes the **exact** same code path as _ErrorStateRecoveredAlt_. With that said, there is a small difference between the three tests and because this class is so critical to how the file-writer works, I would prefer to leave all three tests if (or rather when) we decide to modify the logic implemented here.

Reply:
I usually start reviewing code by looking at the unit tests because they should tell me what the developer intended to implement. I found these two tests misleading because I expected to find a different code path in the implementation. I'm not a fan extra "just in case" tests because they, at best, convolute the documentation.

Reply:
Fair enough, I will agree to removing those two test cases if you think it is ok to remove `Empty` and `TimedOut` from `enum class PollStatus`. They are not actually useful in the current implementation.